### ISSUE TITLE
[4.6.1] Windows Multimedia Keys

### ISSUE THREAD

**Issue by jcursiolf** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#3293]
**OS:** Windows 10 Home 64-bit

**GPMDP Version:** 4.6.1

**Issue Descriptions:** Multimedia track navigation keys misbehaving.

**Steps to Reproduce:** Pressing multimedia key forward on the keyboard will result in two tracks skipped. Pressing multimedia key backward on the keyboard will result in two tracks skipped. Example: assume you're listening to track 4 of a playlist/album. Pressing forward will go to track 6 instead of 5.

**Note:** Similar to closed [#2107](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/2107) from version 4.1.1.

**Comment by Techman-** [comment]
I've come across this issue as well. It's especially prominent when I'm playing a video game (fullscreen).

**Comment by amicloud** [comment]
I am unable to reproduce this issue. The multimedia keys work exactly as expected. On Win 10 64, GPMDP 4.6.1 as well.

**Comment by Klorkson** [comment]
I, too, have this issue on WIndwos 10.
If it matters, my keyboard is the Logitech G910.
Also, in GPMDP (using YouTube Premium Music), I have the Windows 10 System Media Service (with song information) enabled.
What I have also noticed is that sometimes, when skipping songs, the playback gets paused.
It also happens both when the program is open in the fore gorund or in the back ground.

**Comment by jcursiolf** [comment]
Adding a little bit more information: I've reinstalled and it initially did not present the problem. However, I remembered that I like to have the song info on the "Windows 10 Volume Control" and "Windows 10 Media Server". Enabling the Windows 10 Media Server option brought the bug back.  Disabling it made the bug go away again.

**Comment by Techman-** [comment]
Yeah, it seems that disabling the Win10 media service gets rid of the issue, but it also gets rid of the nice volume overlay.

**Comment by Alipoodle** [comment]
Double of: #2364
This is mainly due to Windows Media Service + GPMDP Still both picking up on the Next Media Key it seems.

**Comment by Techman-** [comment]
With the Windows 10 media service disabled in settings, how do I get GPMDP to capture the media key events so I can control it while in a fullscreen game?

**Comment by NetkoNefarious** [comment]
I have this same issue as well.

**Comment by aaldrich29** [comment]
Same issue on 4.6.1 with Media Service enabled, Win 10 Pro 64-bit.

**Comment by JJSax** [comment]
It seems like you may already be aware of this, but at least for me, this same thing is happening with the play/pause button too.  Semi-frequently it will trigger twice, causing my music to stutter but keep playing.

It's consistently happening with both back and forward skipping every time.
These only happen using this application and my keyboard is a Logitech G910 Orion, if it helps.

**Comment by TheAtomicNord** [comment]
This doesn't seem to be getting any attention from @MarshallOfSound but figured I'd add my experience too.
I am having the same issue with Windows 10 64-bit, a logitech g910, and the Windows Media Service enabled much like @Klorkson . Happens regardless of if GPMDP is focused or not.
I like the overlay too much to disable it but it is quite annoying as this was "fixed" in 4.1.2.

**Comment by nickmykkanen** [comment]
Also having this issue. It only happens for the skip forward/skip backward buttons (play/pause seem fine). As others have mentioned, it definitely appears to be the fact that the button is controlling the Windows Media Service _and_ GPMDP, so it receives the skip forward/backward signal twice. Don't think the specific keyboard I have matters, but it's a Logitech MK710.

**Comment by tf2manu994** [comment]
> Looking into it a little more, it seems that the media hotkeys on [keyboardshortcut.js](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/3405b719e39d809f116b841b2b8a651645897de6/src/main/features/core/keyboardShortcuts.js) need to be disabled when the W10 media service is enable, because it gets handled at [win10.js](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/3fa46e76c4b05de51c687d5b1b6272ae129c49ec/src/renderer/windows/GPMWebView/playback/systemMediaService/win10.js).
> 
> [This](https://github.com/timmevandermeer/Google-Play-Music-Desktop-Player-UNOFFICIAL-/commit/e11b54639c493590b7455b593536dfc717ed2585#diff-1a98018c7484c813f86c59d6bf122bc5R15) commit seems like it would fix the issue, but I know that there was some hesitation because not all users have this issue. Is it possible to add a setting for "Media Key Bug Fix", so that users with the issue could choose to enable it?

Can you make a pull request for this?

**Comment by IndefiniteBen** [comment]
>> ...
> > [This](https://github.com/timmevandermeer/Google-Play-Music-Desktop-Player-UNOFFICIAL-/commit/e11b54639c493590b7455b593536dfc717ed2585#diff-1a98018c7484c813f86c59d6bf122bc5R15) commit seems like it would fix the issue, but I know that there was some hesitation because not all users have this issue. Is it possible to add a setting for "Media Key Bug Fix", so that users with the issue could choose to enable it?
> 
> Can you make a pull request for this?

I doubt it. @Kytech previously forked the repo and fixed the issue [here](https://github.com/Kytech/Google-Play-Music-Desktop-Player-UNOFFICIAL-) which was over a year ago and we still have the bug. See his [comment on the issue](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/2364#issuecomment-292858152). Shouldn't this issue be closed for #2364

**Comment by tf2manu994** [comment]
@indefiniteben, so is this a won't fix?

**Comment by IndefiniteBen** [comment]
You'd have to ask @MarshallOfSound I guess (you're a contributor so probably know better than me)? But #2364 has a won't fix from him, but that just means someone else needs to make a complete fix. See this quote from that issue: 

> I don't really have time to go chasing patches 😆 If someone has a fix it would be good to submit as a PR here so I can review and potentially merge to get everyone on the right track.

I take it that if that commit fixes the problem without introducing more issues, if a PR is approved it'll be implemented? But then I thought a fix was about to land a year ago when Kytech made his fix. 

As a user: I take it that this change would not be a "fix" but a toggle between current style and Windows style media control. I don't know if it's acceptable from a dev PoV, but it is from mine.

**Comment by tf2manu994** [comment]
It's acceptable from mine. I've only contributed to documentation, you know
about as much as me for the rest.

**Comment by djh816** [comment]
I actually have an issue where my track skip media keys only work intermittently. Anyone know any fixes? Does this for me even with app in foreground and no change if I have a fullscreen app up (still works, but intermittently). No other media apps seem to have this issue so I can only assume it's something with how this one handles media keys?

**Comment by AlphonseFX** [comment]
This issue still occurs for me when I enable Windows 10 System Media Service in the options. 
Disabling it does fix everything, but as mentioned above, there's no track info without it.

**Comment by Ryltarr** [comment]
I noticed this issue a while ago, and was surprised it hasn't been fixed... I can reliably toggle the issue by turning on/of "Show track information in Windows 10 volume overlay"; when that setting is checked media keys register twice, when that setting is unchecked media keys register once.

I also have a G910, but I tested it with another keyboard (some generic Dell one) and the issue persisted.

**Comment by diskraider** [comment]
At this point I am just adding to the list of people saying they are having the issue. I too am experiencing this.

It seems that if the in app track skip feature was killed when the Windows 10 media setting was enabled it would resolve this issue.

Settings on: Enable the Windows 10 System Media Service, Show track information in Windows 10 volume overlay.

Troubleshooting steps: After disabling the "Show track information in Windows 10 volume overlay" The issue subsided. I noted that the "Enable the Windows 10 System Media Service" is supposed to show info on the lock screen but does not. Issue no longer exists and skipping works fine. Enable that option again and the skipping action is happening twice again.

OS: Windows 10 Pro x64 (1803)

I love the app and the Discord integration. Great work!

**Comment by jonyroda97** [comment]
Same issue here. Skipping twice when "Enable the W10 System Media Service" is checked.

**App Version:** 4.6.1
**OS:** Windows 10 Pro x64 (1809)

**Comment by Drsela** [comment]
Same issue here.

As previously mentioned, clicking Next/Previous track using Media Keys with "Enable the W10 System Media Service" causes GPMP to skip two tracks instead of a single on.

Disabling it causes normal behavior, but shows no "Volume Overlay" due to being disabled.

**App Version:** 4.6.1
**OS:** Windows 10 Pro x64 Version 1809 (OS Build 17763.475)

**Comment by Kytech** [comment]
@IndefiniteBen, I'll be starting up work again on this fix towards the end of the month. I haven't had access to a PC for a while and have been unable to update and test my fix for a more recent version before I submitted the pull request. I'll have my progress updated on #2364 as I make progress.
